### PR TITLE
Add state machine processor, listener & operation

### PR DESCRIPTION
**Pull request by loic425**
| Q               | A
| --------------- | -----
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

### CODE REVIEW

File: src/Component/Metadata/Index.php [hunk 1]
```diff
@@ -36,6 +36,9 @@ public function __construct(
         ?bool $write = null,
         ?string $formType = null,
         ?array $formOptions = null,
+        ?string $stateMachineComponent = null,
+        ?string $stateMachineTransition = null,
+        ?string $stateMachineGraph = null,
```
Comment (lchrusciel) [inline src/Component/Metadata/Index.php]: I'm quite sure, we don't want to support state machine transition on Index operation

File: src/Component/Metadata/HttpOperation.php [hunk 2]
```diff
@@ -34,8 +34,12 @@ public function __construct(
         ?string $grid = null,
         ?bool $read = null,
         ?bool $write = null,
+        ?bool $validate = null,
         ?string $formType = null,
         ?array $formOptions = null,
+        ?string $stateMachineComponent = null,
+        ?string $stateMachineTransition = null,
+        ?string $stateMachineGraph = null,
```
Comment (lchrusciel) [inline src/Component/Metadata/HttpOperation.php]: This should not be generic operation in my opinion. Only creating and updating should take advantage of it

File: src/Component/Metadata/Delete.php [hunk 3]
```diff
@@ -35,6 +35,9 @@ public function __construct(
         ?bool $write = null,
         ?string $formType = null,
         ?array $formOptions = null,
+        ?string $stateMachineComponent = null,
+        ?string $stateMachineTransition = null,
+        ?string $stateMachineGraph = null,
```
Comment (lchrusciel) [inline src/Component/Metadata/Delete.php]: I'm wondering if we support state machine transition on delete. Can you check?

File: src/Component/Metadata/Show.php [hunk 4]
```diff
@@ -36,6 +36,9 @@ public function __construct(
         ?bool $write = null,
         ?string $formType = null,
         ?array $formOptions = null,
+        ?string $stateMachineComponent = null,
+        ?string $stateMachineTransition = null,
+        ?string $stateMachineGraph = null,
```
Comment (lchrusciel) [inline src/Component/Metadata/Show.php]:
```suggestion
```

File: src/Bundle/test/src/Subscription/Entity/Subscription.php [hunk 5]
```diff
@@ -27,23 +29,30 @@
...
     section: 'admin',
+    formType: SubscriptionType::class,
     templatesDir: 'crud',
     routePrefix: '/admin',
 )]
 #[Index(grid: 'app_subscription')]
 #[Create]
 #[Update]
+#[ApplyStateMachineTransition(stateMachineTransition: 'accept')]
+#[ApplyStateMachineTransition(stateMachineTransition: 'reject')]
```
Comment (lchrusciel) [inline src/Bundle/test/src/Subscription/Entity/Subscription.php:40]: Is it enough data to add such an operation? Don't we miss the graph here? How does it work?

File: src/Component/Metadata/Delete.php (hunk 3 above)
Comment (loic425) [reply src/Component/Metadata/Delete.php]: I see no state machine transition usage on deleteAction, but I only see state machine usage on createAction but not on updateAction and that's weird.

File: src/Bundle/test/src/Subscription/Entity/Subscription.php (hunk 5 above)
Comment (loic425) [reply src/Bundle/test/src/Subscription/Entity/Subscription.php:40]: That graph is an optional value. It's only when you use several state machines on a same resource.

File: src/Component/Metadata/HttpOperation.php (hunk 2 above)
Comment (loic425) [reply src/Component/Metadata/HttpOperation.php]: Hum indeed, so maybe with an interface to know if state machine is applicable.

File: src/Bundle/test/src/Subscription/Entity/Subscription.php (hunk 5 above)
Comment (lchrusciel) [reply src/Bundle/test/src/Subscription/Entity/Subscription.php:40]: How do you fetch it, then? Do we expect it to be called default, or do we just fetch the only existing for given class?

File: src/Bundle/test/config/packages/test/sylius_grid.yaml [hunk 6]
```diff
@@ -1,6 +1,7 @@
 sylius_grid:
     templates:
         action:
+            apply_transition: "grid/action/apply_transition.html.twig"
```
Comment (lchrusciel) [inline src/Bundle/test/config/packages/test/sylius_grid.yaml]:
```suggestion
            apply_transition: 'grid/action/apply_transition.html.twig'
```

File: src/Bundle/test/config/packages/test/winzou_state_machine.yaml [hunk 7]
```diff
@@ -32,3 +32,18 @@ winzou_state_machine:
             reopen:
                 from: [ closed ]
                 to: review
+    subscription:
+        class: App\Subscription\Entity\Subscription
+        property_path: state
```
Comment (lchrusciel) [inline src/Bundle/test/config/packages/test/winzou_state_machine.yaml:37]: Minor thing,  but we are mixing `state` and `currentState` variables. Property is called state, but getter/setter are for `currentState`. TBH, I'm surprised this configuration even work, I would expect exception here, as there is no `setState` method on Subscription

File: src/Bundle/test/src/Tests/Controller/SubscriptionUiTest.php [hunk 8]
```diff
@@ -137,6 +137,23 @@ public function it_allows_deleting_a_subscription(): void
         $this->assertEmpty($subscriptions);
     }
 
+    /** @test */
+    public function it_allows_accepting_a_subscription(): void
```
Comment (lchrusciel) [inline src/Bundle/test/src/Tests/Controller/SubscriptionUiTest.php:141]: Is this test executed with Symfony state machine? Do we have the same for winzou?

File: src/Component/Metadata/MetadataInterface.php [hunk 9]
```diff
@@ -13,6 +13,9 @@
 
 namespace Sylius\Component\Resource\Metadata;
 
+/**
+ * @method string|null getStateMachineComponent()
```
Comment (lchrusciel) [inline src/Component/Metadata/MetadataInterface.php:17]: Do we need this soft method requirement?

File: src/Bundle/test/src/Subscription/Entity/Subscription.php (hunk 5 above)
Comment (loic425) [reply src/Bundle/test/src/Subscription/Entity/Subscription.php:40]: On Winzou, the graph is called "default" is you do not specified it =>
 `$graph = $operation->getStateMachineGraph() ?? 'default';`

This is here https://github.com/Sylius/SyliusResourceBundle/pull/612/files#diff-fccfca1acf785d9965e1c335fbcc071e6dffb7433489d6d2341b760ae68e2b4dR36

On Workflow, this is the "workflow name" and is value is null by default
https://github.com/Sylius/SyliusResourceBundle/pull/612/files#diff-52c7ef7ce600296cd01a16a35c3c604f6bc0ab7e8b8179138ad17111cf912267R36

File: src/Bundle/test/config/packages/test/winzou_state_machine.yaml (hunk 7 above)
Comment (loic425) [reply src/Bundle/test/config/packages/test/winzou_state_machine.yaml:37]: Indeed, I think this works cause my property is public. I'm gonna try to change that.

File: src/Bundle/test/src/Tests/Controller/SubscriptionUiTest.php (hunk 8 above)
Comment (loic425) [reply src/Bundle/test/src/Tests/Controller/SubscriptionUiTest.php:141]: I think so, cause we run with the two components on the CI and we have no "specific state machine configuration" for this resource as we can do now.

File: src/Component/Metadata/MetadataInterface.php (hunk 9 above)
Comment (loic425) [reply src/Component/Metadata/MetadataInterface.php:17]: Yeah I think that's better to inform that this method has been introduced.

Comment (loic425) [reply src/Component/Metadata/MetadataInterface.php:17]: Hum that let me think that I have to check if this method exists when I use it on a service.
Hum.. What do you think?

File: src/Bundle/test/src/Subscription/Entity/Subscription.php (hunk 5 above)
Comment (lchrusciel) [reply src/Bundle/test/src/Subscription/Entity/Subscription.php:40]: Exactly, but the name of the graph in winzou is `graph: subscription`. Therefore, it shouldn't work out-of-the-box. I assume that winzou config is not tested. Am I correct?

File: src/Component/Metadata/MetadataInterface.php (hunk 9 above)
Comment (lchrusciel) [reply src/Component/Metadata/MetadataInterface.php:17]: either new interface or always expect this method to be present

**Review by lchrusciel — APPROVED** [review]
Let's merge it, but also add task for refactor of winzou tests to confirm, that it works as expected

**Pull request merged**